src/printer.js

```diff
@@ -109,6 +109,14 @@ function genericPrint(path, options, printPath) {
     needsParens = path.needsParens();
   }
 
+  if (node.type) {
+    // HACK: ASI prevention in no-semi mode relies on knowledge of whether
+    // or not a paren has been inserted (see `exprNeedsASIProtection()`).
+    // For now, we're just passing that information by mutating the AST here,
+    // but it would be nice to find a cleaner way to do this.
+    node.needsParens = needsParens;
+  }
```

Comment:
Cool. I think if we were to make a change at this point I'd rather go to reading tokens. It dodges this issue through elegance rather than complexity, and should be more robust/performant to boot. 

My primary thinking is that it's fundamentally an output-token-based problem, not an AST problem (other than only applying it to statements, which is useful). The question you're trying to answer when deciding whether to prepend a semicolon isn't "is this a TypeCastExpression" or what have you, it's "does this start with a paren". 

As the algorithm stands, we're vulnerable to parens being inserted in the future in places that we're not currently looking for them. It's only through extensive smoke tests that I was able to discover cases like ArrowFunctionExpressions, which sometimes begin with a paren, for example. 

Note that I think the class body stuff should probably still be done with the AST, since the path is more appropriate there. 

... maybe it's time I stop talking and start doing with this, and we can see how it looks. I'll add as a second commit, and will be happy to go back to the AST approach if that's what we prefer.